text/0000-dynamic-attributes.md

```diff
@@ -0,0 +1,189 @@
+
+This proposal allows not just `lwc:dynamic` but all other elements to accept an object which is bound as props/attrs/listeners
+at runtime.
+
+### Prior Art
+- React ([JSX Spread Operator](https://reactjs.org/docs/jsx-in-depth.html#spread-attributes))
+- Vue ([`v-bind`](https://v3.vuejs.org/api/directives.html#v-bind))
+- Svelte ([Spread Props](https://svelte.dev/tutorial/spread-props))
+- Lit ([Discussion for support](https://github.com/lit/lit/issues/923), [PR](https://github.com/lit/lit/pull/1960))
+
+## Detailed Design
+
+A new LWC directive, `lwc:spread` will be introduced that can be used to spread the props/attributes/listeners
+to children elements.
+
+### Structure
+It accepts an object with keys as the attribute/property/listener names and values as the corresponding attribute/property/listener values.
+
+This object is merged with props/attrs/listeners passed in the template (essentially `Object.assign`). However, since template
+compiler splits all the props passed in the template into attributes, properties and event listeners, we do the same thing at runtime. We use the same algorithm that the template compiler uses and whatever the [engine does for native custom elements](https://rfcs.lwc.dev/rfcs/lwc/0000-third-party-web-components-support#proposed-solution-2).
+
+There is one difference however: the set of properties passed in the template doesn't change but it changes in `lwc:spread` i.e.
+an attribute can't be present in one render and absent in the next render if it is defined in the template. With
+`lwc:spread` though, it can be present in one render and not be present in the next.
+
+- attrs/listeners are added/removed based on their presence/absence in `lwc:spread`.
+- properties can't be "removed" if they are not present in subsequent renders. So, we set them to `undefined`. New properties are
+set as usual.
+
+An interesting side effect of the above behavior can be observed when properties have default values:
+
+```html
+// x/app.html
+<template>
+    <x-child lwc:spread={props}></x-child>
+</template>
+```
+```js
+export default class Child extends LightningElement {
+    @api name = "lwc";
+}
+
+export default class App extends LightningElement {
+    props = {};
+    renderedCallback() {
+        // in initial render props is {} and `name` in child is 'lwc'
+        this.props = {name: "aura"}; // `name` in child is "aura";
+        this.props = {}; // `name` in child is `undefined`;
+    }
+}
+
+```
+Customers should be educated on the above behavior.
+
+### Reactivity
+There's no precendent in LWC for change in the set of attributes/properties/listeners. An element can't have a property in one
+render and not have it in another render. LWC Engine actually throws when it detects such a change. In the case of `lwc:spread`,
+we'll have to relax this validation.
+
+In every render, `lwc:spread` will merge with the older value of `lwc:spread`. Any keys not present in the new `lwc:spread` but
```

Comment:
I also desagree with the semantics described here. I don't think we should diff the obj, that's too easy to trick.